File: libs/hayeah-ts/src/config/config.test.ts
import { describe, it, expect, beforeEach, afterEach } from "vitest";
import { load } from "./index.js";
import { readFileSync, writeFileSync, mkdirSync, rmSync } from "fs";
import { join, dirname } from "path";
import { fileURLToPath } from "url";
import { tmpdir } from "os";

const __dirname = dirname(fileURLToPath(import.meta.url));
const testdata = JSON.parse(
  readFileSync(join(__dirname, "../../../testdata/single-envar-config.json"), "utf-8"),
);

const ENV_VAR = "HAYEAH_TEST_CONFIG";
const TMP_DIR = join(tmpdir(), "hayeah-config-test");

beforeEach(() => {
  mkdirSync(TMP_DIR, { recursive: true });
  delete process.env[ENV_VAR];
});

afterEach(() => {
  delete process.env[ENV_VAR];
  rmSync(TMP_DIR, { recursive: true, force: true });
});

describe("config load", () => {
  for (const tc of testdata.load_tests) {
    it(tc.name, () => {
      if (tc.env_value === null || tc.env_value === undefined) {
        delete process.env[ENV_VAR];
      } else if (tc.file_contents !== undefined) {
        // Write a temp file and point the env var at it
        const ext = tc.env_value.endsWith(".toml") ? ".toml" : ".json";
        const filePath = join(TMP_DIR, `config${ext}`);
        writeFileSync(filePath, tc.file_contents);
        process.env[ENV_VAR] = filePath;
      } else if (tc.env_value.startsWith("/nonexistent/")) {
        // Missing file path — use as-is
        process.env[ENV_VAR] = tc.env_value;
      } else {
        process.env[ENV_VAR] = tc.env_value;
      }

      const result = load(ENV_VAR);
      expect(result).toEqual(tc.expected);
    });
  }
});

File: libs/hayeah-go/config/config_test.go
package config

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
)

type loadTest struct {
	Name         string          `json:"name"`
	EnvValue     *string         `json:"env_value"`
	FileContents *string         `json:"file_contents"`
	Expected     json.RawMessage `json:"expected"`
}

type testData struct {
	LoadTests []loadTest `json:"load_tests"`
}

func loadTestData(t *testing.T) testData {
	t.Helper()
	data, err := os.ReadFile("../../testdata/single-envar-config.json")
	if err != nil {
		t.Fatal(err)
	}
	var td testData
	if err := json.Unmarshal(data, &td); err != nil {
		t.Fatal(err)
	}
	return td
}

func TestLoad(t *testing.T) {
	td := loadTestData(t)
	envVar := "HAYEAH_TEST_CONFIG"

	for _, tc := range td.LoadTests {
		t.Run(tc.Name, func(t *testing.T) {
			os.Unsetenv(envVar)
			defer os.Unsetenv(envVar)

			if tc.EnvValue == nil {
				// env var unset
			} else if tc.FileContents != nil {
				// Write a temp file
				tmpDir := t.TempDir()
				var ext string
				if len(*tc.EnvValue) > 5 && (*tc.EnvValue)[len(*tc.EnvValue)-5:] == ".toml" {
					ext = ".toml"
				} else {
					ext = ".json"
				}
				filePath := filepath.Join(tmpDir, "config"+ext)
				if err := os.WriteFile(filePath, []byte(*tc.FileContents), 0o644); err != nil {
					t.Fatal(err)
				}
				os.Setenv(envVar, filePath)
			} else if tc.EnvValue != nil && len(*tc.EnvValue) > 0 && (*tc.EnvValue)[0] == '/' {
				// Missing file path
				os.Setenv(envVar, *tc.EnvValue)
			} else if tc.EnvValue != nil {
				os.Setenv(envVar, *tc.EnvValue)
			}

			result, err := Load(envVar)
			if err != nil {
				t.Fatal(err)
			}

			// Compare by marshaling both to JSON
			gotJSON, _ := json.Marshal(result)
			// Normalize expected
			var expected map[string]any
			if err := json.Unmarshal(tc.Expected, &expected); err != nil {
				t.Fatal(err)
			}
			expectedJSON, _ := json.Marshal(expected)

			if string(gotJSON) != string(expectedJSON) {
				t.Fatalf("got %s, want %s", gotJSON, expectedJSON)
			}
		})
	}
}
